File: Linked-List/linked-list.js
// Create a Node first
class Node {
	constructor(value) {
		this.value = value;
		this.next = null;
	}
}

// LinkList class will use the Node class above to start
class LinkedList {
	constructor(value) {
		const newNode = new Node(value);
		this.head = newNode;
		this.tail = this.head;
		this.length = 1;
	}

	// Push a node into the end of the the linked-list. O(1)
	push(value) {
		const newNode = new Node(value);
        
        // Check if the linked list is empty???
		if (!this.head) {
			this.head = newNode;
			this.tail = newNode;
        } else {
            // Get the tail.next first, why???
			this.tail.next = newNode;
			this.tail = newNode;
		}

		this.length++;
		return this;
	}

	// Pop a node out from the end of the linked list. O(n)
	pop() {
		// Check if the LL is empty
		if (!this.head) return undefined;

		// If length of LL > 1
		let temp = this.head
		let pre = this.head
		while (temp.next) {
			pre = temp;
			temp = temp.next;			
		}
		this.tail = pre;
		this.tail.next = null;
		this.length--

		// If the LL is empty
		if (this.length === 0) {
			this.head = null;
			this.tail = null;
		}
		return temp // temp is the item popped off
	}
}

let firstNode = new LinkedList(4);
firstNode.push(5);
firstNode.push(6);
console.log(firstNode);

File: Linked-List/linked-list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class LinkedList:
    def __init__(self, value):
        self.value = value
        new_node = Node(self.value)

        self.head = new_node
        self.tail = self.head
        self.length = 1

    def push(self, value):
        new_node = Node(value)
        if not(self.head):
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node
        self.length += 1
        return self

    def pop(self):
        if not(self.head):
            return None

        temp = self.head
        pre = self.head

        while temp.next:
            pre = temp
            temp = temp.next

        self.tail = pre
        self.tail.next = None
        self.length -= 1

        if self.length == 0:
            self.head = None
            self.tail = None

        return temp


first_node = LinkedList(4)
print(first_node.value)
print(first_node.head.value)
first_node.push(5)
print('-------')
print(first_node.value)
print(first_node.head.value)
print(first_node.tail.value)
print(first_node.tail.next)
first_node.push(6)
print('-------')
print(first_node.value)
print(first_node.head.value)
print(first_node.head.next.value)
print(first_node.tail.value)
print(first_node.tail.next)
print(first_node.length)
print('===========')
first_node.pop()
print(first_node.value)
print(first_node.head.value)
print(first_node.head.next.value)
print(first_node.tail.value)
print(first_node.tail.next)
print(first_node.length)
